subprojects/docs/src/docs/userguide/authoring-builds/basics/writing_plugins.adoc

```diff
@@ -0,0 +1,207 @@
+                group = "from my plugin"
+                description = "Create myfile.txt in the current directory"
+                fileText.set("HELLO FROM MY PLUGIN")
+            }
+        }
+    }
+}
+
+apply<MyPlugin>()                                                                   // <3>
+----
+<1> Subclass `DefaultTask()`.
+<2> Extend the `org.gradle.api.Plugin` interface.
+<3> Apply the script plugin.
+
+=== Subclass `DefaultTask()`
+First, build a task by subclassing `DefaultTask()`.
+
+[source]
+----
+abstract class CreateFileTask : DefaultTask() { }
+----
+
+This simple task adds a file to our application’s root directory.
+
+=== Extend the `org.gradle.api.Plugin` interface
+Next, create a new class that extends the `org.gradle.api.Plugin` interface.
+
+[source]
+----
+abstract class MyPlugin : Plugin<Project> {
+    override fun apply() {}
+}
+----
+
+You can add tasks and other logic in the `apply()` method.
+
+=== Apply the script plugin
+Finally, apply the plugin in the build script.
+
+[source]
+----
+apply<MyPlugin>()
+----
+
+When `MyPlugin` is applied in the build script, it is calling the method `override fun apply() {}` as defined in the custom `MyPlugin` class.
+
+This makes the plugin available to the application.
+
+NOTE: Script plugins are NOT recommended.
+
+The best practice for developing plugins is to create convention plugins or binary plugins.
+
+== Convention Plugins
+
+Convention plugins are a way to encapsulate and reuse common build configuration in Gradle.
+They allow you to define a set of conventions for a project, and then apply those conventions to other projects or modules.
+
+The example above has been re-written as a convention plugin stored in `buildSrc`:
+
+.buildSrc/src/main/kotlin/MyConventionPlugin.kt
+[source,kotlin]
+----
+import org.gradle.api.DefaultTask
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.OutputFile
+import org.gradle.api.tasks.TaskAction
+import java.io.File
+
+abstract class CreateFileTask : DefaultTask() {
+@get:Input
+abstract val fileText: Property<String>
+
+    @Input
+    val fileName = project.rootDir.toString() + "myfile.txt"
+
+    @OutputFile
+    val myFile: File = File(fileName)
+
+    @TaskAction
+    fun action() {
+        myFile.createNewFile()
+        myFile.writeText(fileText.get())
+    }
+}
+
+class MyConventionPlugin : Plugin<Project> {
+    override fun apply(project: Project) {
+        project.tasks.register("createFileTask", CreateFileTask::class.java) {
+            group = "from my plugin"
+            description = "Create myfile.txt in the current directory"
+            fileText.set("HELLO FROM MY PLUGIN")
+        }
+    }
+}
+----
+
+Which can then be applied in other build scripts accordingly:
+
+.build.gradle.kts
+[source,kotlin]
+----
+import com.gradle.plugin.MyConventionPlugin
+
+apply<MyConventionPlugin>()
```

Comment:
💭 This should probably show the newer `plugins {}` syntax.

Reply:
Can script plugins use the newer syntax?

Reply:
Yes, as long as you define an ID for it.  Much of the existing docs use `apply` so that they can write the `Plugin` class and apply it to the project within a buildscript, for simplicity.  If you've already split the plugin out to it's own file in `buildSrc` it probably makes sense to show a complete example of defining the id and using the new syntax.  See https://dev.to/rogervinas/first-steps-developing-custom-gradle-plugins-4m51#create-plugins-in-the-buildsrc-module for an example.